frontend/src/types/schemaSystem.ts

```diff
@@ -0,0 +1,351 @@
+import { SchemaReviewPolicyId } from "./id";
+import { Principal } from "./principal";
+
+export enum RuleLevel {
+  DISABLED = "DISABLED",
+  ERROR = "ERROR",
+  WARNING = "WARNING",
+}
+
+export const LEVEL_LIST = [
+  RuleLevel.ERROR,
+  RuleLevel.WARNING,
+  RuleLevel.DISABLED,
+];
+
+export enum PayloadType {
+  STRING = "STRING",
+  STRING_ARRAY = "STRING[]",
+  TEMPLATE = "TEMPLATE",
+}
+
+interface StringPayload {
+  type: PayloadType.STRING;
+  default: string;
+  value?: string;
+}
+
+interface StringArrayPayload {
+  type: PayloadType.STRING_ARRAY;
+  default: string[];
+  value?: string[];
+}
+
+interface TemplatePayload {
+  type: PayloadType.TEMPLATE;
+  default: string;
+  templateList: { id: string; description?: string }[];
+  value?: string;
+}
+
+interface NamingFormatRulePayload {
+  format: StringPayload | TemplatePayload;
+}
+
+interface RequiredColumnsRulePayload {
+  list: StringArrayPayload;
+}
+
+export type RulePayload =
+  | NamingFormatRulePayload
+  | RequiredColumnsRulePayload
+  | undefined;
+
+export type SchemaRuleEngineType = "MYSQL" | "COMMON";
+
+export type CategoryType = "ENGINE" | "NAMING" | "QUERY" | "TABLE" | "COLUMN";
+
+export interface SchemaRule {
+  id: string;
+  category: CategoryType;
+  engine: SchemaRuleEngineType;
+  description: string;
```

Comment:
If we look at our database schema and the other places, the normal way to do this is we declare a type field in the SchemaRule, and having a payload type which is a composite of NamingFormatRulePayload | RequiredColumnsRulePayload | StringArrayPayload etc...

The code written would be

switch (type) {
case STRING:
...
case STRING_ARRAY:
...
}

Reply:
The payload could contain many configurations so we cannot simply use one payload type for one rule, that's why I'm using an object for the payload.

For example, a rule could be

```json
{
  "id": "naming.column",
  "category": "NAMING",
  "engine": "COMMON",
  "description": "",
  "payload": {
    "format": {
      "type": PayloadType.STRING,
      "default": "^[a-z]+(_[a-z]+)?$",
    },
    "length": {
      "type": PayloadType.NUMBER,
      "default": 20
    }
  }
}
```

Even though we don't have such rule for now, but seems it's possible to support this at any time in the future

Reply:
We have a similar pattern for our issue/task type: https://sourcegraph.com/github.com/bytebase/bytebase/-/blob/api/task.go?L85&subtree=true

We can try to generalize, but ok to create a separate payload if different. The difference from the original design is we do NOT allow users to create customized rules. Thus all rules are managed by us, thus makes it possible to apply more strict typing and prevent unexpected dirty data.

Reply:
I'm confused about what we're talking about. The `type` for the payload is different with the `type` for the rule (or task). If we want to put the type in the rule, we should define different rule types instead of payload types, and remove the payload type:

```ts
interface BaseRuleType {
  id: RuleId;
  category: CategoryType;
  engine: EngineType;
  level: RuleLevel;
}

interface NamingFormatRuleType extends BaseRuleType {
  format: string;
}

interface NamingFormatWithLengthLimitRuleType extends BaseRuleType {
  format: string;
  length: number;
}

interface RequiredColumnRuleType extends BaseRuleType {
  columnList: string[];
}

type DatabaseSchemaRule = BaseRuleType | NamingFormatRuleType | NamingFormatWithLengthLimitRuleType | RequiredColumnRuleType;

interface DatabaseSchemaReviewPolicy {
  id: SchemaReviewPolicyId;

  // Standard fields
  creator: Principal;
  createdTs: number;
  updater: Principal;
  updatedTs: number;

  // Domain specific fields
  name: string;
  ruleList: DatabaseSchemaRule[];
  environmentIdList: number[];
}
```

This could be less flexible, but I'm ok with it for use in the API and the backend.

Reply:
Update, now we will use `RuleTemplate` for the frontend, and `SchemaPolicyRule` for the backend and the API.
The `SchemaPolicyRule` contains different rule types, the backend can use the rule id to parse the real rule struct.